Add `withDescription` for naming MR steps

aposwolsky opened the pull request:
Adding ability to have more meaningful names in MR steps instead of just "(X/Y")

johnynek commented:
This is pretty great. Thanks for doing this. I'll review it carefully this week.

aposwolsky commented:
will change.  any other issues? @johnynek ?

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -31,6 +36,25 @@ trait ExecutionContext {
 
+  def getIdentifierOpt(descriptions: Seq[String]): Option[String] = {
+    if (descriptions.nonEmpty) {
+      Some(descriptions.distinct.mkString(", "))
```
can we sort these to make sure it is stable?

johnynek replied:
let's make this a private method on `object ExecutionContext`.

aposwolsky replied:
agreed to make it private.  I wanted the sorting to be in order and did the distinct just to be safe.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -31,6 +36,25 @@ trait ExecutionContext {
+  def updateFlowStepName(step: BaseFlowStep[JobConf], descriptions: Seq[String]): BaseFlowStep[JobConf] = {
+    getIdentifierOpt(descriptions).foreach(newIdentifier => {
+      val stepXofYData = """\(\d+/\d+\)""".r.findFirstIn(step.getName).getOrElse("")
+      // Reflection is only temporary.  Latest cascading has setName public: https://github.com/cwensel/cascading/commit/487a6e9ef#diff-0feab84bc8832b2a39312dbd208e3e69L175
```
add an issue to the issue tracker so we don't lose track of this, and link the issue here.

aposwolsky replied:
I'll also add it in the comment.
https://github.com/twitter/scalding/issues/1294

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -52,6 +76,25 @@ trait ExecutionContext {
+          val descriptions: Seq[String] = {
+            for {
+              vertex <- baseFlowStep.getGraph.vertexSet.toList
+              if (vertex.isInstanceOf[Pipe])
```
```scala
def getDesc(v: Any): Seq[String] = v match {
  case vPipe: Pipe => RichPipe.getPipeDescriptions(vPipe)
  case _ => List() // no descriptions
}
```
then call that rather than the test + cast (do it in one go and no explicit cast).

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -52,6 +76,25 @@ trait ExecutionContext {
       }
       val flow = mode.newFlowConnector(withId).connect(flowDef)
 
+      if (flow.isInstanceOf[HadoopFlow]) {
```
```scala
flow match {
  case hflow: HadoopFlow =>
    // logic here
  case _ =>
    // descriptions not yet supported in other modes
}
```

aposwolsky replied:
looks good! will change

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -31,6 +36,25 @@ trait ExecutionContext {
+    } else {
+      None
+    }
+  }
```
nit: our style is to only use `{ }` when needed (more than one expression). Many of your `{ }` are redundant and we'd prefer if you not add them.

aposwolsky replied:
sounds good

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala:
```diff
@@ -52,6 +53,38 @@ object RichPipe extends java.io.Serializable {
     p
   }
+
+  // A pipe can have more than one description when merged together, so we store them delimited witg 254.toChar.
```
typo: with

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala:
```diff
@@ -52,6 +53,38 @@ object RichPipe extends java.io.Serializable {
   }
+
+  // A pipe can have more than one description when merged together, so we store them delimited witg 254.toChar.
+  def encodePipeDescriptions(descriptions: Seq[String]): String = {
```
make private as much as possible to make it easier to preserve binary compatibility.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala:
```diff
@@ -423,11 +440,15 @@ case class ValueSortedReduce[K, V1, V2](
+
+  override def withDescription(description: String) =
+    ValueSortedReduce[K, V1, V2](
+      keyOrdering, mapped, valueSort, reduceFn, reducers, descriptions ++ Seq(description))
```
`descriptions :+ description` would be better.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala:
```diff
@@ -469,6 +492,9 @@ case class IteratorMappedReduce[K, V1, V2](
     copy(reducers = Some(red))
 
+  override def withDescription(description: String): IteratorMappedReduce[K, V1, V2] =
+    copy(descriptions = descriptions ++ Seq(description))
```
`use :+ rather than ++`

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/typed/WithDescription.scala:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Used for objects that may _set_ a description to be used in .dot and MR step names.
+ */
+trait WithDescription[+This <: WithDescription[This]] extends HasDescription {
```
why must it have `HasDescription`? without that, we could add this to `TypedPipe[T]` which I think would be useful.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala:
```diff
@@ -210,6 +211,9 @@ trait TypedPipe[+T] extends Serializable {
   def debug: TypedPipe[T] = onRawSingle(_.debug)
 
+  /** adds a description to the pipe */
+  def withDescription(description: String): TypedPipe[T] = new WithDescriptionTypedPipe[T](this, description)
```
should add `TypedPipe[T] extends WithDescription[TypedPipe[T]]`, don't you think?

aposwolsky replied:
I liked this approach because it just wrapped a pipe with a description, and you can see how easy it was.  otherwise every version of TypedPipe is going to have to set the description in their ".toPipe" ... this puts it all in one place and it didn't seem as we lose any functionality...  it is analogous to the way we are doing WithOnComplete on TypedPipe

johnynek replied:
Sounds fine to me. Just wanted to see your thoughts here. I don't really see us writing a method like:

```scala
def describe[T <: WithDescription[T]](t: T, s: String): T = t.withDescription(s)
```
which is the only think I can think of us doing with this.

So, let's leave it as is.

ianoc replied:
The other PR going in now puts a common toPipe wrapper in the TypedPipe trait for setting some state on our pipes before handing them out. When that merges doing the TypedPipe extends... will be more straight forward since all typed pipe instances are just suppling asPipe

johnynek commented:
Have you tried this on your cluster at all? It's hard to be sure we are hitting all the cases (hashJoin, joins, groupBys, forceToDisk are all different code paths). Could we beef up the tests at the very least to try each of those cases?

aposwolsky commented:
Yes tried it on a bunch of our jobs which used all those cases (except for
hashJoin perhaps).  I'll add more tests!

On Thu, May 14, 2015 at 4:33 PM, P. Oscar Boykin <notifications@github.com>
wrote:

> Have you tried this on your cluster at all? It's hard to be sure we are
> hitting all the cases (hashJoin, joins, groupBys, forceToDisk are all
> different code paths). Could we beef up the tests at the very least to try
> each of those cases?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/scalding/pull/1283#issuecomment-102160266>.
>

johnynek commented:
Making a macro to get the current line and file would be useful with this:

`.withDescription(fileAndLine)` could be

ianoc commented on scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala:
```diff
@@ -52,6 +53,38 @@ object RichPipe extends java.io.Serializable {
+  }
+
+  def setPipeDescriptions(p: Pipe, descriptions: Seq[String]): Pipe = {
+    p.getStepConfigDef().setProperty(
```
This will mostly work, but... our OrderedSerialization issues seem to have popped up that these won't always go together :/  -- we tried to use these to configure serializers for each step. Do an op that would do some transform then added the config to that pipe to let it work. But then the two would be moved apart by cascading. Not a blocker to this, but it unfortunately would mean the naming ends up being more best effort than would be ideal I think

aposwolsky replied:
I see your point, and I wonder if we remove "withDescription" from typedPipe and only allow it after you ".group" if that would guarantee better that it is attached to the step where it occurs ... but then it is also less powerful and it is nice to annotate descriptions everywhere

ianoc replied:
Funnily enough after a group by is exactly where we've been seeing it fail :/ I'm not sure its worth altering the plan here. This is a big step up in info/usability. We might just want to include a little note about YMMV i guess in a readme

ianoc commented:
Looks like this has legit failures against latest develop now, doesn't compile

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala:
```diff
@@ -1036,6 +1040,19 @@ class WithOnComplete[T](typedPipe: TypedPipe[T], fn: () => Unit) extends TypedPi
 }
 
+class WithDescriptionTypedPipe[T](typedPipe: TypedPipe[T], description: String) extends TypedPipe[T] {
+  override def toPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]) = {
```
this should be asPipe now.

ianoc commented:
@aposwolsky we are aiming to freeze for a rls on wednesday. It would be awesome to get this in for that if we can

ianoc commented:
@aposwolsky we are looking to freeze for our next rls at 5pm pst today, you going to have a chance to look at this before then?

aposwolsky commented:
I was OOO the past few days, but I rebased and pushed the change from toPipe to asPipe

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -31,6 +36,38 @@ trait ExecutionContext {
+    val conf = step.getConfig
+    getIdentifierOpt(getDesc(step)).foreach(id => {
+      val newJobName = "%s %s".format(conf.getJobName, id)
+      println("Added descriptions to job name: %s".format(newJobName))
```
can we get a log.info or debug here instead of a println?

johnynek replied:
also, this looks like it is changing how all jobs are named. That is somewhat orthogonal to descriptions, right?

We have a lot of tooling that assumes the job names are not changed. Could we instead just add a key to the jobConf?

aposwolsky replied:
hmmm, we actually like it to show it in hadoop jobtracker, what about we make it configurable so we will keep it setting name and we can default to something else?

aposwolsky replied:
also, the "job name" always had "(X/Y)" appending to it, all this is doing is appending extra stuff after that, not changing the prefix.

johnynek replied:
We have some analytics that run on our jobs and they match on the job name. For instance, we have a reducer estimator that runs and looks at the names to get the history of the particular step. Others might have other systems running.

I think changing how the jobs are names could be disruptive for people.

Can we find a way to opt-in to this behavior and you can put those options as default in your clusters? Something like a configuration key like `"scalding.description.addtoname" -> "true"` or something? When this is not present, we just add an entry to the jobConf?

aposwolsky replied:
I already pushed a change that would put the descriptions in a different config and not update mapred.name.  Since we want it appended to the name I will add a steplistener on our side that will just add the value to that config to the job name on our side... win win for everyone :)

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala:
```diff
@@ -53,6 +54,38 @@ object RichPipe extends java.io.Serializable {
+
+  // A pipe can have more than one description when merged together, so we store them delimited with 254.toChar.
+  private def encodePipeDescriptions(descriptions: Seq[String]): String = {
+    descriptions.map(_.replace(254.toChar, ' ')).filter(_.nonEmpty).mkString(254.toChar.toString)
```
can we factor `254.toChar` into a private val? Also, can you comment on the choice here? What impact does this have for UTF-8 descriptions? A guess since means that the 254 code point:  http://www.fileformat.info/info/unicode/char/fe/index.htm can't appear in descriptions. Why not `1.toChar` which is often used as a delimiter on HDFS?

aposwolsky replied:
1.toChar is just as good, I can switch to that.

aposwolsky commented:
Made suggested changes

aposwolsky commented:
how do we re-run the tests, it seems to have failed on spurious errors:
15/05/28 22:09:48 ERROR mapred.TaskTracker: Caught exception: java.io.IOException: Call to localhost/127.0.0.1:35898 failed on local exception: java.nio.channels.ClosedByInterruptException

ianoc commented:
@aposwolsky i triggered a rebuild of your branch, not sure you can do it yourself :/

aposwolsky commented:
What's up with these errors? Is this pull request ready to be merged?

ianoc commented:
Looks like legit failures, same target fails in 2.10 and 2.11 -- 

[error] Failed tests:
[error] 	com.twitter.scalding.platform.WithDescriptionTest
[error] (scalding-hadoop-test/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 1079 s, completed May 29, 2015 6:19:38 PM

aposwolsky commented:
Thanks, the log didn't show any interesting details but it showed up when I ran the failing test locally. I had a bug not changing everything from char 254 to char 1, but it also turned out that using char 1 as a delimiter caused issues, so I switched it to 255 and pushed fix.

aposwolsky commented:
Can you trigger it to run again, the failure is now: "java.lang.OutOfMemoryError: PermGen space"

ianoc commented:
rebuilding..

aposwolsky commented:
yay it worked!

aposwolsky commented:
ping?

johnynek commented:
looking

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -52,6 +87,18 @@ trait ExecutionContext {
+          flowSteps.foreach(step => {
+            val baseFlowStep: BaseFlowStep[JobConf] = step.asInstanceOf[BaseFlowStep[JobConf]]
+            val descriptions = getDesc(baseFlowStep)
+            updateFlowStepName(baseFlowStep, descriptions)
```
looks like you are still updating the job name. Like I said, I don't want to change that without more thought and buy in. Our existing job monitoring will be impacted and probably other systems will as well. I know we should have a test for anything we care about, and we can add an issue to do this here (if indeed we want some contract on the name and not another way to track job identity).

aposwolsky replied:
The mapred.name is already set by the time it gets here, so the job.xml file for hadoop will not have any mention of the descriptions. However, the "stepName" needs to be updated for the dot file generation to have the descriptions

aposwolsky commented:
@johnynek : ok, I changed it so it only changed the flow stepName in the .dot file generation, otherwise it just sets a different config. This works for our setup since (1) .dot files have descriptions and (2) I added a flowstepstrategy that adds the descriptions to the job name so it looks good on our side.

Does this solution work for you as well? I would like to see this pull request go through because I don't want to have to start maintaining a patch.

aposwolsky commented:
ping?

aposwolsky commented:
ping?

johnynek commented:
Looking now. Sorry.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ExecutionContext.scala:
```diff
@@ -31,6 +36,24 @@ trait ExecutionContext {
+    if (descriptions.nonEmpty) Some(descriptions.distinct.mkString(", ")) else None
+  }
+
+  private def updateStepConfigWithDescriptions(step: BaseFlowStep[JobConf], descriptions: Seq[String]): Unit = {
```
descriptions is unused, right?

aposwolsky replied:
good catch! thanks will fix

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala:
```diff
@@ -53,6 +54,39 @@ object RichPipe extends java.io.Serializable {
+    else {
+      val encodedResult = p.getStepConfigDef.apply(Config.PipeDescriptions, new Getter {
+        override def update(s: String, s1: String): String = ???
+        override def get(s: String): String = null
```
can you comment on what is going on here?

aposwolsky replied:
yep, will add note.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala:
```diff
@@ -53,6 +54,39 @@ object RichPipe extends java.io.Serializable {
+        override def update(s: String, s1: String): String = ???
+        override def get(s: String): String = null
+      })
+      Some(encodedResult)
```
Option(encodedResult)
 .filterNot(_.isEmpty)

might be simpler. But no biggie if you don't want to do this.

aposwolsky replied:
oh good idea, will change.

johnynek commented:
One minor issue, then in looks good to merge (remove an unused parameter). Thanks for pushing this through to the end.

aposwolsky commented:
no problem! made changes

johnynek commented:
Thanks! :+1: Will merge when it's green.

aposwolsky commented:
:( can we re-run the CI build?

johnynek commented:
restarted

aposwolsky commented:
success!

johnynek commented:
Thanks for this nice new feature!

aposwolsky commented:
anytime :)

joshualande commented:
@aposwolsky, thanks for adding this feature. It is really great!
